src/Build/Graph/ProjectGraph.cs

```diff
@@ -22,21 +26,56 @@ internal sealed class ProjectGraph
         /// <exception cref="InvalidProjectFileException">If the evaluation of any project in the graph fails.</exception>
         public ProjectGraph(string entryProjectFile)
         {
-            var graphNode =
-                new ProjectGraphNode(ProjectCollection.GlobalProjectCollection.LoadProject(entryProjectFile));
-            _projectNodes.Add(graphNode);
+            LoadGraph(entryProjectFile);
         }
 
         /// <summary>
-        /// Constructs a graph starting from the given project files.
+        /// Get an unordered collection of all project nodes in the graph.
         /// </summary>
-        /// <param name="entryProjectFiles">The project files to use as the entry points in constructing the graph</param>
-        /// <exception cref="InvalidProjectFileException">If the evaluation of any project in the graph fails.</exception>
-        public ProjectGraph(IEnumerable<string> entryProjectFiles) { }
+        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => _allParsedProjects.Values;
+
+        private ProjectGraphNode CreateNewNode(string projectFilePath)
+        {
+            var graphNode =
+                new ProjectGraphNode(ProjectCollection.GlobalProjectCollection.LoadProject(projectFilePath));
+            _allParsedProjects.Add(projectFilePath, graphNode);
+            return graphNode;
+        }
 
         /// <summary>
-        /// Get an unordered collection of all project nodes in the graph.
+        /// Load a graph with root node at entryProjectFile
+        /// Maintain a queue of projects to be processed- each queue item is a key value pair of the project to be evaluated and its parent
+        /// Once the project has been evaluated, add a project reference to this evaluated target from the parent node
         /// </summary>
-        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => _projectNodes;
+        /// <param name="entryProjectFile"></param>
+        private void LoadGraph(string entryProjectFile)
+        {
+            var projectsToEvaluate = new Queue<KeyValuePair<string, ProjectGraphNode>>();
+            // entry project node has no parent
+            projectsToEvaluate.Enqueue(new KeyValuePair<string, ProjectGraphNode>(entryProjectFile, null));
+            while (projectsToEvaluate.Count != 0)
+            {
+                KeyValuePair<string, ProjectGraphNode> projectToEvaluateAndParentNode = projectsToEvaluate.Dequeue();
+                string projectToEvaluate = projectToEvaluateAndParentNode.Key;
+                if (!_allParsedProjects.TryGetValue(projectToEvaluate, out ProjectGraphNode parsedProject))
+                {
+                    parsedProject = CreateNewNode(projectToEvaluate);
+                    IEnumerable<ProjectItem> projectReferenceItems =
```

Comment:
I thought a node is a tuple of project path and global properties. Is the plan to add global properties later on?